db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,325 @@
+	newRunInit := func() error {
+		uniqueUUID, err := uuid.NewRandom()
+		if err != nil {
+			return err
+		}
+
+		a.MigrationID = uniqueUUID.String()
+		base.InfofCtx(ctx, base.KeyAll, "Attachment Migration: Starting new migration run with migration ID: %s", a.MigrationID)
+		return nil
+	}
+
+	if clusterStatus != nil {
+		var statusDoc AttachmentMigrationManagerStatusDoc
+		err := base.JSONUnmarshal(clusterStatus, &statusDoc)
+
+		// If the previous run completed, or there was an error during unmarshalling the status we will start the
+		// process from scratch with a new migration ID. Otherwise, we should resume with the migration ID, stats specified in the doc.
+		if statusDoc.State == BackgroundProcessStateCompleted || err != nil {
+			return newRunInit()
+		}
+		a.MigrationID = statusDoc.MigrationID
+		a.SetStatus(statusDoc.DocsChanged, statusDoc.DocsProcessed)
+		a.SetCollectionIDs(statusDoc.CollectionIDs)
+
+		base.InfofCtx(ctx, base.KeyAll, "Attachment Migration: Resuming migration with migration ID: %s, %d already processed", a.MigrationID, a.DocsProcessed.Value())
+
+		return nil
+	}
+
+	return newRunInit()
+}
+
+func (a *AttachmentMigrationManager) Run(ctx context.Context, options map[string]interface{}, persistClusterStatusCallback updateStatusCallbackFunc, terminator *base.SafeTerminator) error {
+	db := a.databaseCtx
+	migrationLoggingID := "Migration: " + a.MigrationID
+
+	persistClusterStatus := func() {
+		err := persistClusterStatusCallback(ctx)
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to persist latest cluster status for attachment migration: %v", migrationLoggingID, err)
+		}
+	}
+	defer persistClusterStatus()
+
+	var processFailure error
+	failProcess := func(err error, format string, args ...interface{}) bool {
+		processFailure = err
+		terminator.Close()
+		base.WarnfCtx(ctx, format, args...)
+		return false
+	}
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		docID := string(event.Key)
+		collection := db.CollectionByID[event.CollectionID]
+		base.TracefCtx(ctx, base.KeyAll, "[%s] Received DCP event %d for doc %v", migrationLoggingID, event.Opcode, base.UD(docID))
+
+		// Ignore documents without xattrs, to avoid processing unnecessary documents
+		if event.DataType&base.MemcachedDataTypeXattr == 0 {
+			return true
+		}
+
+		// Don't want to process raw binary docs
+		// The binary check should suffice but for additional safety also check for empty bodies
+		if event.DataType == base.MemcachedDataTypeRaw || len(event.Value) == 0 {
+			return true
+		}
+
+		// We only want to process full docs. Not any sync docs.
+		if strings.HasPrefix(docID, base.SyncDocPrefix) {
+			return true
+		}
+
+		a.DocsProcessed.Add(1)
+		syncData, _, _, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, collection.userXattrKey(), false)
+		if err != nil {
+			failProcess(err, "error unmarshaling document %s with migrationID %s: %v", base.UD(docID), a.MigrationID, err)
```

Comment:
```suggestion
			failProcess(err, "[%s] error unmarshaling document %s: %v, stopping attachment migration.", migrationLoggingID, base.UD(docID), err)
```

If this does fail, we want to be able to easily grep and find the error message with like messages.